### PR TITLE
Changes for EOS-22165 Implement event manager interface to list (get_events) and get message type APIs

### PR DESCRIPTION
**Pull request by ArchanaLimaye**
## Problem Statement
<pre>
  <code>
  Story Ref (if any):
    Need to implement the functions get_events, message_type
  </code>
</pre>
## Unit testing on RPM done
<pre>
  <code>
  No , testing done on deployed setup using python files directly
  </code>
</pre>
## Problem Description
<pre>
  <code>
    Implement the functions get_events, message_type
  </code>
</pre>
## Solution
<pre>
  <code>
    Implemented functionality for get_events, message_type
  </code>
</pre>
## Unit Test Cases
<pre>
  <code>
   On a 3 node deployment  ddded keys to consul directly.
   Tested by executing the event_manager.py file directly that 
  For both function calls
  -  Correct value returned when matching key is found
  -  None / empty list returned when exact matching key is not found
  -  All success / failure paths in code tested ( single key / multiple keys with same prefix / no matching keys present in consul etc)
  </code>
</pre>

### CODE REVIEW

File: ha/core/event_manager/event_manager.py [hunk 1]
```diff
@@ -332,4 +345,16 @@ def message_type(self, component: str) -> str:
         Args:
             component (str): component name.
         """
+        key = const.EVENT_MGR_MESSAGE_TYPE_KEY.replace("<component_id>", component)
+        value = None
+        Log.debug(f"Fetching message type for {key}")
+
```
Comment (ajay-paratmandali) [inline ha/core/event_manager/event_manager.py:351]: kv = self._confstore.get(key)
if kv is None:
      return kv
else:
      ....

File: ha/core/event_manager/event_manager.py [hunk 2]
```diff
@@ -332,4 +345,16 @@ def message_type(self, component: str) -> str:
...
+        if self._confstore.key_exists(key):
+            kv = self._confstore.get(key)
```
Comment (ajay-paratmandali) [inline ha/core/event_manager/event_manager.py:353]: for loop not need as key will return only one key val pair 

is single component will have multiple message type

File: ha/core/event_manager/event_manager.py [hunk 3]
```diff
@@ -318,6 +318,19 @@ def get_events(self, component : str) -> list:
...
 
+        key = f'{const.COMPONENT_KEY}/{component}'
+        value = []
+        Log.debug(f"Fetching subscribed events for {key}")
+
+        if self._confstore.key_exists(key):
+            kv = self._confstore.get(key)
+            for k, v in kv.items():
+                if k.endswith(key):
+                    value = json.loads(v)
```
Comment (ajay-paratmandali) [inline ha/core/event_manager/event_manager.py:329]: value = []
```
value = json.loads(v)
```
assigned again is it append

can we use 
return kv.values()

**Review by ajay-paratmandali — CHANGES_REQUESTED** [review]

File: ha/core/event_manager/event_manager.py (hunk 2 above)
Comment (ArchanaLimaye) [reply ha/core/event_manager/event_manager.py:353]: There can be multiple keys with same prefix (not expected but keys can get added my mistake) and we want the exact matching key ex   "events/subscribe/sspl"  ,  "events/subscribe/sspl_test" etc
So added the for loop

File: ha/core/event_manager/event_manager.py (hunk 3 above)
Comment (ArchanaLimaye) [reply ha/core/event_manager/event_manager.py:329]: kv.values returns a string and that needs to be converted

File: ha/core/event_manager/event_manager.py (hunk 1 above)
Comment (ArchanaLimaye) [reply ha/core/event_manager/event_manager.py:351]: Done

File: ha/core/event_manager/event_manager.py [hunk 4]
```diff
@@ -318,6 +318,19 @@ def get_events(self, component : str) -> list:
             list: List of events.
         """
 
+        key = f'{const.COMPONENT_KEY}/{component}'
```
Comment (ajay-paratmandali) [inline ha/core/event_manager/event_manager.py:321]: key = f'{const.COMPONENT_KEY}/{component}'

for all key use _prepare_key()
all keys should be in one place to make it easy to maintain

File: ha/core/event_manager/event_manager.py [hunk 5]
```diff
@@ -332,4 +345,16 @@ def message_type(self, component: str) -> str:
         Args:
             component (str): component name.
         """
+        key = const.EVENT_MGR_MESSAGE_TYPE_KEY.replace("<component_id>", component)
```
Comment (ajay-paratmandali) [inline ha/core/event_manager/event_manager.py:348]:
```
def publish(self, component:str, event: RecoveryActionEvent) -> None:
```
return type not matching
i think publish job is to send event in message bus

File: ha/core/event_manager/event_manager.py (hunk 3 above)
Comment (ajay-paratmandali) [reply ha/core/event_manager/event_manager.py:329]:
```
def get_events(self, component : str) -> list:
```
and
```
value = json.loads(v)
```
both have different return type

File: ha/core/event_manager/event_manager.py [hunk 6]
```diff
@@ -318,6 +318,19 @@ def get_events(self, component : str) -> list:
             list: List of events.
         """
 
+        key = f'{const.COMPONENT_KEY}/{component}'
+        value = []
+        Log.debug(f"Fetching subscribed events for {key}")
+
+        kv = self._confstore.get(key)
+        if kv:
+            for k, v in kv.items():
```
Comment (ajay-paratmandali) [inline ha/core/event_manager/event_manager.py:327]: In both function do we need for loop 

are they returning multiple value

File: ha/core/event_manager/event_manager.py (hunk 5 above)
Comment (ArchanaLimaye) [reply ha/core/event_manager/event_manager.py:348]: this is message_type function , not publish

File: ha/core/event_manager/event_manager.py (hunk 4 above)
Comment (ArchanaLimaye) [reply ha/core/event_manager/event_manager.py:321]: key creation here is based on the # defines done in earlier PRs.
We can have a generic key creation for all functions , not just get_events and message_type

That can be done in this PR or separately @ajay-paratmandali @indrajitzagade any suggestions  ?

File: ha/core/event_manager/event_manager.py (hunk 6 above)
Comment (ArchanaLimaye) [reply ha/core/event_manager/event_manager.py:327]: addressing case where keys with same prefix might get added 
ex: "message_type/sspl  , "message_type/sspl_test"  etc
OR the case where passed component string is incomplete ex:  function call looks like message_type("har")
       This should not return the list meant for "hare"

File: ha/core/event_manager/event_manager.py (hunk 3 above)
Comment (ArchanaLimaye) [reply ha/core/event_manager/event_manager.py:329]: Lets discuss, I am not clear

File: ha/core/event_manager/event_manager.py (hunk 6 above)
Comment (ArchanaLimaye) [reply ha/core/event_manager/event_manager.py:327]: _confstore.get can return multiple KVs ; where  "key" is the matching prefix. So for loop is needed

File: ha/core/event_manager/event_manager.py (hunk 4 above)
Comment (indrajitzagade) [reply ha/core/event_manager/event_manager.py:321]: Yes, I agree. we need generic key creation.

**Review by ajay-paratmandali — APPROVED** [review]

**Review by indrajitzagade — APPROVED** [review]

**Comment by indrajitzagade** [comment]
@ArchanaLimaye  Please resolve conflict.

**Comment by ArchanaLimaye** [comment]
conflicts are resolved

**Pull request merged**